Specs/Scene/ModelExperimental/StatisticsPipelineStageSpec.js

```diff
@@ -0,0 +1,491 @@
+      const components = gltfLoader.components;
+      const primitive = components.nodes[1].primitives[0];
+      const positionAttribute = primitive.attributes[1];
+
+      const mockRuntimePrimitive = {
+        positionBuffer2D: {
+          sizeInBytes: positionAttribute.typedArray.byteLength,
+        },
+      };
+
+      // Simulate how the 2D stage unlinks the typed array. In practice, this
+      // array may still be referenced by the loader so the statistics stage
+      // will count it.
+      positionAttribute.typedArray = undefined;
+
+      StatisticsPipelineStage._count2DPositions(
+        statistics,
+        mockRuntimePrimitive
+      );
+
+      // This stage intentionally counts the GPU + CPU copies of the 2D
+      // positions
+      const positions2D = mockRuntimePrimitive.positionBuffer2D;
+      expect(statistics.geometryByteLength).toBe(2 * positions2D.sizeInBytes);
+    });
+  });
+
+  it("_countMorphTargetAttributes updates memory statistics", function () {
+    return loadGltf(animatedMorphCube).then(function (gltfLoader) {
+      const statistics = new ModelExperimentalStatistics();
+      const components = gltfLoader.components;
+      const primitive = components.nodes[0].primitives[0];
+
+      StatisticsPipelineStage._countMorphTargetAttributes(
+        statistics,
+        primitive
+      );
+
+      let totalSize = 0;
+      const morphTargets = primitive.morphTargets;
+      const morphTargetsLength = morphTargets.length;
+      for (let i = 0; i < morphTargetsLength; i++) {
+        const attributes = morphTargets[i].attributes;
+        const attributesLength = attributes.length;
+        for (let j = 0; j < attributesLength; j++) {
+          totalSize += attributes[j].buffer.sizeInBytes;
+        }
+      }
+
+      expect(statistics.geometryByteLength).toBe(totalSize);
+    });
+  });
+
+  it("_countMaterialTextures updates memory statistics for metallic roughness", function () {
+    return loadGltf(boomBox).then(function (gltfLoader) {
+      const components = gltfLoader.components;
+      const primitive = components.nodes[0].primitives[0];
+
+      const material = primitive.material;
+      const metallicRoughness = material.metallicRoughness;
+      const statistics = new ModelExperimentalStatistics();
+
+      StatisticsPipelineStage._countMaterialTextures(
+        statistics,
+        primitive.material
+      );
+
+      const totalTextureSize =
+        material.emissiveTexture.texture.sizeInBytes +
+        material.normalTexture.texture.sizeInBytes +
+        material.occlusionTexture.texture.sizeInBytes +
+        metallicRoughness.baseColorTexture.texture.sizeInBytes;
+      // metallic roughness texture is the same image file as the base
+      // color texture so it is not counted
+      expect(statistics.texturesByteLength).toBe(totalTextureSize);
+    });
+  });
+
+  it("_countMaterialTextures updates memory statistics for specular glossiness", function () {
```

Comment:
Should we test that `_countMaterialTextures` doesn't update for a model without material textures? (e.g. simple texture model, or simple model without a material?)

Also I don't think `getAllTextureReaders` accounts for if a model doesn't have a material

Reply:
I tried this with `Triangle.gltf` and discovered that it still has a default material! Though surprisingly, it wasn't any GltfLoader code, it was https://github.com/CesiumGS/gltf-pipeline this time! I left some comments explaining this, because it was non-obvious